app/models/service/aggregation.rb

```diff
@@ -9,12 +9,121 @@ module Service::Aggregation
+      # else
+      #   aggregation_sql = Hardware.arel_table[hardware_column].sum
+      # end
+
+      # aggregation_sql = Hardware.arel_table[hardware_column]
+      # aggregation_sql = aggregation_sql * 1.megabyte if options[:mb_to_bytes]
+      # aggregation_sql = Arel::Nodes::NamedFunction.new('NUMERIC', [aggregation_sql.sum])
+
+      aggregation_sql = Hardware.arel_table[hardware_column].sum
+      aggregate_hardware_arel(virtual_column_name, aggregation_sql, options)
+    end
+
+    def aggregate_all_vms_disk_count_arel
+      virtual_column_name = "aggregate_all_vm_disk_count"
+      aggregation_sql = Disk.arel_table[:id].count
+      aggregate_hardware_arel(virtual_column_name, aggregation_sql, :include_disks => true)
+    end
+
+    def aggregate_all_vms_disk_space_allocated_arel
+      zero            = Arel::Nodes::SqlLiteral.new("0")
+      column_name     = "aggregate_all_vm_disk_space_allocated"
+      coalesce_values = [ Disk.arel_table[:size], zero ]
+      aggregation_sql = Arel::Nodes::NamedFunction.new('SUM',
+        [Arel::Nodes::NamedFunction.new('COALESCE', coalesce_values)]
+      )
+      aggregate_hardware_arel(column_name, aggregation_sql, :include_disks => true)
+    end
+
+    def aggregate_all_vms_disk_space_used_arel
+      zero            = Arel::Nodes::SqlLiteral.new("0")
+      column_name     = "aggregate_all_vm_disk_space_used"
+      coalesce_values = [ Disk.arel_table[:size_on_disk], Disk.arel_table[:size], zero ]
+      aggregation_sql = Arel::Nodes::NamedFunction.new('SUM',
+        [Arel::Nodes::NamedFunction.new('COALESCE', coalesce_values)]
+      )
+      aggregate_hardware_arel(column_name, aggregation_sql, :include_disks => true)
+    end
+
+    def aggregate_hardware_arel(virtual_column_name, aggregation_sql, options={})
+      lambda do |t|
+        subtree_services             = Arel::Table.new(:services)
+        subtree_services.table_alias = "#{virtual_column_name}_services"
+
+        service_resources   = ServiceResource.arel_table
+        vms                 = Vm.arel_table
+        hardwares           = Hardware.arel_table
+
+        vm_or_template_type = Arel::Nodes::SqlLiteral.new("'VmOrTemplate'")
+        ancestry_match      = Arel::Nodes::SqlLiteral.new("'/%'")
+        ancestry_ilike      = Arel::Nodes::NamedFunction.new("CONCAT", [t[:id], ancestry_match])
+        service_id_cast     = Arel::Nodes::NamedFunction.new("CAST", [t[:id].as("VARCHAR")])
+
+        where_clause   = t.grouping(
+          subtree_services[:id].eq(t[:id])
+           .or(subtree_services[:ancestry].matches(ancestry_ilike))
+        ).or(subtree_services[:ancestry].eq(service_id_cast))
+
+        vm_join_clause = vms[:id].eq(service_resources[:resource_id])
+        if options[:on_vms_only] # meaning VMs that are powerd "on"
+          vm_join_clause = vm_join_clause.and(vms[:power_state].lower.eq('on'))
+        end
+
+        subselect =
+          subtree_services.project(aggregation_sql)
```

Comment:
So that is kind of what I was getting at (poorly) with this (again, quoting PR description):

> * Optimally, sharing the `JOIN`s here would be ideal as they would prevent making multiple loops in the database for each aggregate column. I was originally under the impression that the current UI doesn't facilitate for something like that easily, but it might be possible based on how I have implemented the scopes for this and the changes to the controller. This would be a minor improvement at this point, and the `virtual_column` arel improvements might still want to be included regardless.

Essentially, we could have the SQL be something more like this instead:

```SQL
SELECT id,
       SUM(hardwares.cpu_total_cores) AS aggregate_all_vm_cpus
       SUM(hardwares.memory_mb) AS aggregate_all_vm_memory
FROM services
JOIN service_resources ON service_resources.service_id = services.id
                       AND service_resources.resource_type = 'VmOrTemplate' 
JOIN vms               ON vms.id = service_resources.resource_id
JOIN hardwares         ON hardwares.vm_or_template_id = vms.id
WHERE (("services"."id" = services.id 
               OR "services"."ancestry" ILIKE CONCAT(services.id, '/%'))
              OR "services"."ancestry" = CAST(services.id AS VARCHAR)))
WHERE id = <my_service.id>;
```

And just have the scope `with_aggregates` include those columns and joins.  When I was originally working on this, I was under the impression that using the `virtual_attribute` pattern was going be required as this was pulled from a `MiqReport`, and only around Friday did I realize the record actually displaying these aggregates was not going through the report.  I had all this code worked up, so I wanted to put it up for review as is, but it might be that this isn't the best way to go about it.

"Hashtage #PipeDream":  Given the time and bandwidth to do this completely differently, I would argue that pulling in the stats as part of the main request is something we should not do, and we rub some of the "MSA Special Sauce" (Mirco Services Architecture) on this and make a separate request to retrieve these aggregate stats, and avoid slowing down the main query.